doc/serial-timeout-troubleshoot.md

```diff
@@ -0,0 +1,80 @@
+# Troubleshooting "serial timeout" errors
+
+Several nRF Connect Desktop applications use serial ports to communicate with
+development kits and dongles. nRF Connect Desktop also uses serial ports to
+perform DFU (Device Firmware Update) when a devkit/dongle is running a 
+[DFU bootloader](http://infocenter.nordicsemi.com/topic/com.nordic.infocenter.sdk5.v15.0.0/sdk_app_serial_dfu_bootloader.html?cp=4_0_0_4_3_4).
+
+Serial ports might be [UARTs](http://infocenter.nordicsemi.com/topic/com.nordic.infocenter.nrf52840.ps/uart.html?cp=2_0_0_5_32)
+on the devkits connected to a RS-232 port on a computer,
+or they might be emulated through USB CDC ACM (also known as "Virtual COM port",
+"serial over USB" or other names). The later happens when using 
+[USB ACM CDC on the IMCU](http://infocenter.nordicsemi.com/topic/com.nordic.infocenter.nrf52/dita/nrf52/development/preview_dev_kit/vir_com_port.html?cp=2_1_5_4_1),
+or when using [USB functionality from the Nordic SDK](http://infocenter.nordicsemi.com/topic/com.nordic.infocenter.sdk5.v15.0.0/lib_usbd_class_cdc.html?cp=4_0_0_3_51_8_3)
+on SoCs that implement USB (e.g. the nRF52840).
+
+While serial ports are ubiquitous, serial port transport lacks any protocol
+metadata or capabilities for request-response higher-level protocols. Combined
+with the fact that modern desktop operating systems still allow for legacy
+serial ports, this can trigger timeouts.
+
+This document collects some known scenarios and workarounds for these timeout 
+errors.
+
+# Scenarios where these errors happen
+
+
+## DevKit/dongle not running bootloader
+
+When using a nRF52840 devkit/dongle and performing a DFU operation with 
+nRF Connect Desktop, a `Timeout while reading from serial transport.` error shows up.
+
+When using a nRF52840 devkit/dongle and performing a DFU operation with
+[nrfutil](https://github.com/NordicSemiconductor/pc-nrfutil/), a 
+`TypeError: 'NoneType' object is not iterable` error shows up.
```

Comment:
This looks like a bug. Looks like there is a missing null check somewhere in the Python code.

I tried to reproduce this by doing a DFU with nrfutil when the nRF52840 dongle was running the connectivity application. I got a different error; a traceback ending with:

```
  File "/usr/local/lib/python2.7/dist-packages/nordicsemi/dfu/dfu_transport_serial.py", line 206, in open
    if self.__ping() == True:
  File "/usr/local/lib/python2.7/dist-packages/nordicsemi/dfu/dfu_transport_serial.py", line 327, in __ping
    if resp[0] != DfuTransportSerial.OP_CODE['Response']:
IndexError: list index out of range
```

Maybe we should be a bit more generic here? F.ex. "error messages like ... could show up".

Meanwhile, I will create a ticket for improving the error messages in this scenario. In the ticket I will also add a comment about updating this documentation again when a proper error message is in place.

Reply:
That's the error message *I* got when doing a DFU. I kinda expected the python implementation to give out the same error message when the response from a DFU target is empty/unparseable, seems I was wrong.